### PR TITLE
Pubsub node RDBMS backend

### PR DESCRIPTION
**Pull request by michalwski**
This PR adds RDBMS backend for `pubsub_node` table.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[5934.1](https://travis-ci.org/esl/MongooseIM/jobs/460935024) / Erlang 19.3 / small_tests / 7d6ff83239d6b3c1689a27bc5f209e665883499f
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2145/5934/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2145/5934/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-28_19.27.58/lib.mongooseim.logs/run.2018-11-28_19.27.59/suite.log.html)

**Comment by mongoose-im** [comment]
[5935.1](https://travis-ci.org/esl/MongooseIM/jobs/461179609) / Erlang 19.3 / small_tests / a56a4d5023378497fcc24dee6cc8d3ae550c32f3
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2145/5935/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2145/5935/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-29_09.13.43/lib.mongooseim.logs/run.2018-11-29_09.13.44/suite.log.html)

File: priv/mssql2012.sql [hunk 1]
```diff
@@ -474,6 +474,31 @@ GO
...
 GO
 
+CREATE TABLE dbo.pubsub_nodes (
+    nidx BIGINT           NOT NULL,
+    p_key NVARCHAR(250)   NOT NULL,
+    name NVARCHAR(250)    NOT NULL,
+    type NVARCHAR(250)    NOT NULL,
+    owners NVARCHAR(max)  NOT NULL,
+    options NVARCHAR(max) NOT NULL,
+    CONSTRAINT PK_pubsub_nodes PRIMARY KEY CLUSTERED(
```
Comment (fenek) [inline priv/mssql2012.sql]: With such order in primary key (applies to all schemas), queries using only `p_key` and/or `name` won't use the index. I think what we need here is `p_key, name, nidx` with possibly separate index for `nidx` alone.

File: priv/mssql2012.sql [hunk 2]
```diff
@@ -474,6 +474,31 @@ GO
 CREATE INDEX i_inbox_ts ON inbox(luser, lserver, timestamp);
 GO
 
+CREATE TABLE dbo.pubsub_nodes (
+    nidx BIGINT           NOT NULL,
+    p_key NVARCHAR(250)   NOT NULL,
+    name NVARCHAR(250)    NOT NULL,
+    type NVARCHAR(250)    NOT NULL,
+    owners NVARCHAR(max)  NOT NULL,
```
Comment (fenek) [inline priv/mssql2012.sql]: Perhaps could be replaced with a JOIN with `pubsub_affiliations`?

File: src/pubsub/mod_pubsub_db_rdbms.erl [hunk 3]
```diff
@@ -228,49 +228,172 @@ del_node(Nidx) ->
...
-    mod_pubsub_db_mnesia:get_parentnodes_tree(Key, Node).
+    find_nodes_with_parents(Key, [Node], 0, []).
+
+find_nodes_with_parents(_, [], _, Acc) ->
+    Acc;
+find_nodes_with_parents(_, _, 100, Acc) ->
+    ?WARNING_MSG("event=max_depth_reached, nodes=~p", [Acc]),
+    Acc;
+find_nodes_with_parents(Key, Nodes, Depth, Acc) ->
+    SQL = mod_pubsub_db_rdbms_sql:select_nodes_by_key_and_names_in_list_with_parents(
```
Comment (fenek) [inline src/pubsub/mod_pubsub_db_rdbms.erl]: What are we going to do one day when a proper function name + module name exceeds 100 characters alone? :D

File: src/pubsub/mod_pubsub_db_rdbms.erl [hunk 4]
```diff
@@ -228,49 +228,172 @@ del_node(Nidx) ->
...
+    NewAcc = [{Depth, NewNodes} | Acc],
+    find_nodes_with_parents(Key, lists:flatten(Parents), Depth + 1, NewAcc).
+
+build_map([], Map) ->
+    Map;
+build_map([Row | Rows], Map) ->
+    NewMap = update_nodes_map(Row, Map),
+    build_map(Rows, NewMap).
+
+update_nodes_map({NidxSQL, NodeName, NextLeveNodeName}, Map) ->
```
Comment (fenek) [inline src/pubsub/mod_pubsub_db_rdbms.erl]: `NextLevelNodeName`

File: src/pubsub/mod_pubsub_db_rdbms.erl [hunk 5]
```diff
@@ -228,49 +228,172 @@ del_node(Nidx) ->
...
+                                                 parents = Parents},
+                             NewNodes = [Node | NodesAcc],
+                             {NewParents, NewNodes}
+                     end,
+
+    {Parents, NewNodes} = maps:fold(MapTransformer, {[], []}, Map),
+    NewAcc = [{Depth, NewNodes} | Acc],
+    find_nodes_with_parents(Key, lists:flatten(Parents), Depth + 1, NewAcc).
+
+build_map([], Map) ->
```
Comment (fenek) [inline src/pubsub/mod_pubsub_db_rdbms.erl]: Kind of `lists:foldl` reimplementation. :)

File: priv/mssql2012.sql (hunk 1 above)
Comment (michalwski) [reply priv/mssql2012.sql]: Good catch!

File: src/pubsub/mod_pubsub_db_rdbms.erl (hunk 3 above)
Comment (michalwski) [reply src/pubsub/mod_pubsub_db_rdbms.erl]: I don't know yet :)

File: src/pubsub/mod_pubsub_db_rdbms.erl (hunk 5 above)
Comment (michalwski) [reply src/pubsub/mod_pubsub_db_rdbms.erl]: Right :) I'll use `lists:foldl` then :)

**Comment by mongoose-im** [comment]
[5949.1](https://travis-ci.org/esl/MongooseIM/jobs/461792402) / Erlang 19.3 / small_tests / 03f2149027c302e7e1e0bf41a03592b588ab1456
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2145/5949/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2145/5949/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_14.48.45/lib.mongooseim.logs/run.2018-11-30_14.48.46/suite.log.html)

**Comment by mongoose-im** [comment]
[5950.1](https://travis-ci.org/esl/MongooseIM/jobs/461870199) / Erlang 19.3 / small_tests / c25a5c73ae721e4abe3ef9c41623dc247031e7a9
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2145/5950/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2145/5950/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_16.52.49/lib.mongooseim.logs/run.2018-11-30_16.52.50/suite.log.html)

**Comment by mongoose-im** [comment]
[5951.1](https://travis-ci.org/esl/MongooseIM/jobs/461907205) / Erlang 19.3 / small_tests / ac68ca6da12e96df65a5ba8fb2ef0ba02f88c1fa
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2145/5951/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2145/5951/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_18.27.29/lib.mongooseim.logs/run.2018-11-30_18.27.29/suite.log.html)

**Pull request merged**